backend/database-sqlite.js

```diff
@@ -0,0 +1,106 @@
+'use strict'
+
+import { mkdir } from 'node:fs/promises'
+import { join, resolve } from 'node:path'
+import sqlite3 from 'sqlite3'
+
+import { checkKey } from './database.js'
+
+let db: any = null
+let readStatement: any = null
+let writeStatement: any = null
+
+const run = (sql, args) => {
+  return new Promise((resolve, reject) => {
+    db.run(sql, args, (err) => {
+      if (err) {
+        reject(err)
+      } else {
+        resolve()
+      }
+    })
+  })
+}
+
+export async function initStorage (options: Object = {}): Promise<void> {
+  const { dirname, filename } = options
+  const dataFolder = resolve(dirname)
+
+  await mkdir(dataFolder, { mode: 0o750, recursive: true })
+
+  await new Promise((resolve, reject) => {
+    if (db) {
+      reject(new Error(`The ${filename} SQLite database is already open.`))
+    }
+    db = new sqlite3.Database(join(dataFolder, filename), (err) => {
+      if (err) {
+        reject(err)
+      } else {
+        resolve()
+      }
+    })
+  })
+  await run('CREATE TABLE IF NOT EXISTS Data(key TEXT NOT NULL PRIMARY KEY, value TEXT NOT NULL)')
+  console.log('Connected to the %s SQLite database.', filename)
+  readStatement = db.prepare('SELECT value FROM Data WHERE key = ?')
+  writeStatement = db.prepare('REPLACE INTO Data(key, value) VALUES(?, ?)')
+}
+
+// Useful in test hooks.
+export function clear (): Promise<void> {
+  return run('DELETE FROM Data')
+}
+
+export function exportToJSON (): Promise<Object> {
+  return new Promise((resolve, reject) => {
+    db.all('SELECT key, value FROM Data', [], (err, rows) => {
```

Comment:
This would read the entire database contents into memory, right? That seems like an extreme amount of data to read into memory (potentially hundreds of gigabytes worth). What is this function for?

Reply:
Copying a note from @snowteamer from Slack about what this is:

> These are database migration helpers, which I was planning to maybe use either in the preloading code or  in chel migrate.
 I've left them here so that they can be unit-tested in database.test.js and reviewed, but yeah actually they are not used (at least yet).